# Deriving New Variables
## Contents: Limits, removing incorrect values

In [1]:
# Import Libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Import DF

orders_prods_merge = pd.read_pickle(r'C:\Users\thoms\Desktop\Data Analytics\Achievement 4\Instacart Basket Analysis\02 Data\Prepared Data\orders_products_combined_1\orders_products_combined_1.pkl')

In [3]:
path = r'C:\Users\thoms\Desktop\Data Analytics\Achievement 4\Instacart Basket Analysis'

In [5]:
# Limit DF to 1000000 rows of data

df = orders_prods_merge[:1000000]

In [6]:
#Check shape

df.shape

(1000000, 15)

In [7]:
#Overview of Data

df.describe()


order_id         user_id    order_number       order_dow  \
count  1.000000e+06  1000000.000000  1000000.000000  1000000.000000   
mean   1.712242e+06     3271.262332       17.286436        2.765079   
std    9.871742e+05     1853.883895       17.899691        2.091724   
min    8.000000e+00        1.000000        1.000000        0.000000   
25%    8.618970e+05     1682.000000        5.000000        1.000000   
50%    1.711867e+06     3319.000000       11.000000        3.000000   
75%    2.564123e+06     4892.000000       24.000000        5.000000   
max    3.420853e+06     6433.000000      100.000000        6.000000   

       order_hour_of_day  days_since_prior_order     product_id  \
count     1000000.000000           936719.000000  993568.000000   
mean           13.418602               11.105355   25575.568444   
std             4.222888                8.838321   14090.214777   
min             0.000000                0.000000       1.000000   
25%            10.000000                5.000000   13517.000000   
50%            13.000000                8.000000   25305.000000   
75%            16.000000               15.000000   37795.000000   
max            23.000000               30.000000   49687.000000   

       add_to_cart_order      reordered       aisle_id  department_id  \
count      993568.000000  993568.000000  992622.000000  992622.000000   
mean            8.307632       0.590916      71.043512       9.983426   
std             7.206951       0.491665      38.228879       6.295089   
min             1.000000       0.000000       1.000000       1.000000   
25%             3.000000       0.000000      31.000000       4.000000   
50%             6.000000       1.000000      83.000000       9.000000   
75%            11.000000       1.000000     107.000000      16.000000   
max            95.000000       1.000000     134.000000      21.000000   

              prices  
count  992622.000000  
mean       11.422743  
std       427.574331  
min         1.000000  
25%         4.200000  
50%         7.400000  
75%        11.300000  
max     99999.000000

In [8]:
#Show all prices with price of 99999.00

df.loc[(df['prices'] == 99999)]

order_id  user_id eval_set  order_number  order_dow  \
129909    183964      873    prior             3          0   
129922   1851256      873    prior             4          6   
286436   1915696     1893    prior             1          5   
286447   2763293     1893    prior             2          4   
286470   2564805     1893    prior             4          1   
502898    949012     3339    prior             1          1   
502941    420057     3339    prior             2          0   
502963   1511997     3339    prior             3          1   
502980   1153448     3339    prior             4          4   
503002   2669259     3339    prior             5          4   
531488    965345     3531    prior            27          2   
773188   2291230     5034    prior             6          1   
773202   2324832     5034    prior             7          6   
960327    323856     6175    prior            37          2   
967639    842052     6222    prior            10          1   

        order_hour_of_day  days_since_prior_order  product_id  \
129909                 10                     7.0     33664.0   
129922                 12                    13.0     33664.0   
286436                 17                     NaN     33664.0   
286447                 16                    13.0     33664.0   
286470                 17                    30.0     33664.0   
502898                 14                     NaN     33664.0   
502941                 11                    13.0     33664.0   
502963                 18                     8.0     33664.0   
502980                 16                    10.0     33664.0   
503002                 14                    21.0     33664.0   
531488                 19                     8.0     33664.0   
773188                 17                    30.0     33664.0   
773202                 11                    30.0     33664.0   
960327                  9                     6.0     33664.0   
967639                 18                    17.0     33664.0   

        add_to_cart_order  reordered           product_name  aisle_id  \
129909               11.0        0.0  2 % Reduced Fat  Milk      84.0   
129922                8.0        1.0  2 % Reduced Fat  Milk      84.0   
286436               10.0        0.0  2 % Reduced Fat  Milk      84.0   
286447                6.0        1.0  2 % Reduced Fat  Milk      84.0   
286470                3.0        1.0  2 % Reduced Fat  Milk      84.0   
502898                8.0        0.0  2 % Reduced Fat  Milk      84.0   
502941               29.0        1.0  2 % Reduced Fat  Milk      84.0   
502963               14.0        1.0  2 % Reduced Fat  Milk      84.0   
502980                1.0        1.0  2 % Reduced Fat  Milk      84.0   
503002                6.0        1.0  2 % Reduced Fat  Milk      84.0   
531488                2.0        0.0  2 % Reduced Fat  Milk      84.0   
773188                3.0        0.0  2 % Reduced Fat  Milk      84.0   
773202                9.0        1.0  2 % Reduced Fat  Milk      84.0   
960327                6.0        0.0  2 % Reduced Fat  Milk      84.0   
967639                2.0        0.0  2 % Reduced Fat  Milk      84.0   

        department_id   prices _merge  
129909           16.0  99999.0   both  
129922           16.0  99999.0   both  
286436           16.0  99999.0   both  
286447           16.0  99999.0   both  
286470           16.0  99999.0   both  
502898           16.0  99999.0   both  
502941           16.0  99999.0   both  
502963           16.0  99999.0   both  
502980           16.0  99999.0   both  
503002           16.0  99999.0   both  
531488           16.0  99999.0   both  
773188           16.0  99999.0   both  
773202           16.0  99999.0   both  
960327           16.0  99999.0   both  
967639           16.0  99999.0   both

In [9]:
# Incorrect price for Reduced milk. Recommend removing all this Data

df = df.drop(labels=[967639, 960327, 773202, 773188, 531488, 503002, 502980, 502963, 502898, 502941, 286470, 286447, 286436, 129922, 129909 ], axis=0)

In [10]:
#Overview of Data - after removal of extreme prices

df.describe()

order_id        user_id   order_number      order_dow  \
count  9.999850e+05  999985.000000  999985.000000  999985.000000   
mean   1.712245e+06    3271.261285      17.286579       2.765082   
std    9.871757e+05    1853.886836      17.899746       2.091725   
min    8.000000e+00       1.000000       1.000000       0.000000   
25%    8.618970e+05    1682.000000       5.000000       1.000000   
50%    1.711867e+06    3319.000000      11.000000       3.000000   
75%    2.564123e+06    4892.000000      24.000000       5.000000   
max    3.420853e+06    6433.000000     100.000000       6.000000   

       order_hour_of_day  days_since_prior_order     product_id  \
count      999985.000000           936706.000000  993553.000000   
mean           13.418584               11.105289   25575.446330   
std             4.222900                8.838305   14090.286089   
min             0.000000                0.000000       1.000000   
25%            10.000000                5.000000   13517.000000   
50%            13.000000                8.000000   25305.000000   
75%            16.000000               15.000000   37795.000000   
max            23.000000               30.000000   49687.000000   

       add_to_cart_order      reordered       aisle_id  department_id  \
count      993553.000000  993553.000000  992607.000000  992607.000000   
mean            8.307638       0.590917      71.043316       9.983335   
std             7.206958       0.491665      38.229135       6.295093   
min             1.000000       0.000000       1.000000       1.000000   
25%             3.000000       0.000000      31.000000       4.000000   
50%             6.000000       1.000000      83.000000       9.000000   
75%            11.000000       1.000000     107.000000      16.000000   
max            95.000000       1.000000     134.000000      21.000000   

              prices  
count  992607.000000  
mean        9.911759  
std       178.158554  
min         1.000000  
25%         4.200000  
50%         7.400000  
75%        11.300000  
max     14900.000000

In [11]:
#check values of 14900 as this is not normal a normal groccery store

df.loc[(df['prices'] == 14900)]

order_id  user_id eval_set  order_number  order_dow  \
1592      912404       17    prior            12          2   
1654      603376       17    prior            22          6   
16668    3264360      135    prior             2          2   
16674     892534      135    prior             3          0   
54053     229704      342    prior             8          1   
...          ...      ...      ...           ...        ...   
966094    717264     6217    prior            12          0   
966099    840159     6217    prior            13          1   
966145   1235750     6217    prior            20          5   
967403     40712     6221    prior             5          0   
967477   2567369     6221    prior            10          2   

        order_hour_of_day  days_since_prior_order  product_id  \
1592                   14                     5.0     21553.0   
1654                   16                     4.0     21553.0   
16668                  21                    13.0     21553.0   
16674                   8                    12.0     21553.0   
54053                  19                    30.0     21553.0   
...                   ...                     ...         ...   
966094                 15                     6.0     21553.0   
966099                 11                    15.0     21553.0   
966145                  9                    11.0     21553.0   
967403                 15                     8.0     21553.0   
967477                 17                    17.0     21553.0   

        add_to_cart_order  reordered                      product_name  \
1592                  5.0        0.0  Lowfat 2% Milkfat Cottage Cheese   
1654                  3.0        1.0  Lowfat 2% Milkfat Cottage Cheese   
16668                 6.0        0.0  Lowfat 2% Milkfat Cottage Cheese   
16674                 3.0        1.0  Lowfat 2% Milkfat Cottage Cheese   
54053                 9.0        0.0  Lowfat 2% Milkfat Cottage Cheese   
...                   ...        ...                               ...   
966094                6.0        1.0  Lowfat 2% Milkfat Cottage Cheese   
966099                4.0        1.0  Lowfat 2% Milkfat Cottage Cheese   
966145                6.0        1.0  Lowfat 2% Milkfat Cottage Cheese   
967403                8.0        0.0  Lowfat 2% Milkfat Cottage Cheese   
967477                3.0        1.0  Lowfat 2% Milkfat Cottage Cheese   

        aisle_id  department_id   prices _merge  
1592       108.0           16.0  14900.0   both  
1654       108.0           16.0  14900.0   both  
16668      108.0           16.0  14900.0   both  
16674      108.0           16.0  14900.0   both  
54053      108.0           16.0  14900.0   both  
...          ...            ...      ...    ...  
966094     108.0           16.0  14900.0   both  
966099     108.0           16.0  14900.0   both  
966145     108.0           16.0  14900.0   both  
967403     108.0           16.0  14900.0   both  
967477     108.0           16.0  14900.0   both  

[142 rows x 15 columns]

# All 2% milk products have incorrect value. Recommendation is to remove these values for now. 142 rows will not effect the analysis

In [12]:
# removed rows with prioduct_id of 21553 as price is incorrect

df = df.drop(df.index[df['product_id'] == 21553.0 ])

In [13]:
df.describe()

order_id        user_id   order_number      order_dow  \
count  9.998430e+05  999843.000000  999843.000000  999843.000000   
mean   1.712252e+06    3271.205347      17.287324       2.765082   
std    9.871696e+05    1853.851397      17.900312       2.091739   
min    8.000000e+00       1.000000       1.000000       0.000000   
25%    8.618970e+05    1682.000000       5.000000       1.000000   
50%    1.711867e+06    3318.000000      11.000000       3.000000   
75%    2.564106e+06    4892.000000      24.000000       5.000000   
max    3.420853e+06    6433.000000     100.000000       6.000000   

       order_hour_of_day  days_since_prior_order     product_id  \
count      999843.000000           936572.000000  993411.000000   
mean           13.418573               11.105032   25576.021306   
std             4.222931                8.838264   14091.211023   
min             0.000000                0.000000       1.000000   
25%            10.000000                5.000000   13517.000000   
50%            13.000000                8.000000   25309.000000   
75%            16.000000               15.000000   37807.000000   
max            23.000000               30.000000   49687.000000   

       add_to_cart_order      reordered       aisle_id  department_id  \
count      993411.000000  993411.000000  992465.000000  992465.000000   
mean            8.307467       0.590917      71.038029       9.982474   
std             7.206928       0.491665      38.229314       6.295132   
min             1.000000       0.000000       1.000000       1.000000   
25%             3.000000       0.000000      31.000000       4.000000   
50%             6.000000       1.000000      83.000000       9.000000   
75%            11.000000       1.000000     107.000000      16.000000   
max            95.000000       1.000000     134.000000      21.000000   

              prices  
count  992465.000000  
mean        7.781313  
std         4.233232  
min         1.000000  
25%         4.200000  
50%         7.300000  
75%        11.300000  
max        25.000000

In [14]:
# define price ranges

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [15]:
df['price_range'] = df.apply(price_label, axis=1)

In [16]:
df['price_range'].value_counts (dropna = False)

Mid-range product    668114
Low-range product    312168
High range            12183
Not enough data        7378
Name: price_range, dtype: int64

In [17]:
# My max price is not the same as in the exercise. I have checked everyline of code since the start of this achievement and can not find the mistake. AS per the exercise then I have made a miscalucaltion or error. This would be good to understand where. 

df['prices'].max()

25.0

In [18]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [19]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [20]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [21]:
df['price_range_loc'].value_counts(dropna = False)

Mid-range product     668114
Low-range product     312168
High-range product     12183
NaN                     7378
Name: price_range_loc, dtype: int64

In [23]:
orders_prods_merge.loc[orders_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [24]:
orders_prods_merge.loc[(orders_prods_merge['prices'] <= 15) & (orders_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [25]:
orders_prods_merge.loc[orders_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [26]:
orders_prods_merge['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
NaN                     236409
Name: price_range_loc, dtype: int64

# Now the Data line up but I still have 236409 lines of NaN data. I am not sure what needs to be done here or if that needs to be dropped. 

In [27]:
orders_prods_merge.isnull().sum()

order_id                        0
user_id                         0
eval_set                        0
order_number                    0
order_dow                       0
order_hour_of_day               0
days_since_prior_order    2078102
product_id                 206209
add_to_cart_order          206209
reordered                  206209
product_name               236409
aisle_id                   236409
department_id              236409
prices                     236409
_merge                          0
price_range_loc            236409
dtype: int64

In [28]:
df = orders_prods_merge [orders_prods_merge['product_name'].isnull() == True]

In [29]:
df

order_id  user_id eval_set  order_number  order_dow  \
59         1187899        1    train            11          4   
255        1492625        2    train            15          1   
344        2774568        3     test            13          5   
363         329954        4     test             6          3   
401        2196797        5    train             5          0   
...            ...      ...      ...           ...        ...   
32639949   1716008   206205    train             4          1   
32640235   1043943   206206     test            68          0   
32640459   2821651   206207     test            17          2   
32641137    803273   206208     test            50          5   
32641267    272231   206209    train            14          6   

          order_hour_of_day  days_since_prior_order  product_id  \
59                        8                    14.0         NaN   
255                      11                    30.0         NaN   
344                      15                    11.0         NaN   
363                      12                    30.0         NaN   
401                      11                     6.0         NaN   
...                     ...                     ...         ...   
32639949                 16                    10.0         NaN   
32640235                 20                     0.0         NaN   
32640459                 13                    14.0         NaN   
32641137                 11                     4.0         NaN   
32641267                 14                    30.0         NaN   

          add_to_cart_order  reordered product_name  aisle_id  department_id  \
59                      NaN        NaN          NaN       NaN            NaN   
255                     NaN        NaN          NaN       NaN            NaN   
344                     NaN        NaN          NaN       NaN            NaN   
363                     NaN        NaN          NaN       NaN            NaN   
401                     NaN        NaN          NaN       NaN            NaN   
...                     ...        ...          ...       ...            ...   
32639949                NaN        NaN          NaN       NaN            NaN   
32640235                NaN        NaN          NaN       NaN            NaN   
32640459                NaN        NaN          NaN       NaN            NaN   
32641137                NaN        NaN          NaN       NaN            NaN   
32641267                NaN        NaN          NaN       NaN            NaN   

          prices     _merge price_range_loc  
59           NaN  left_only             NaN  
255          NaN  left_only             NaN  
344          NaN  left_only             NaN  
363          NaN  left_only             NaN  
401          NaN  left_only             NaN  
...          ...        ...             ...  
32639949     NaN  left_only             NaN  
32640235     NaN  left_only             NaN  
32640459     NaN  left_only             NaN  
32641137     NaN  left_only             NaN  
32641267     NaN  left_only             NaN  

[236409 rows x 16 columns]

In [30]:
# drop all lines with prices as null

orders_prods_merge = orders_prods_merge.drop(orders_prods_merge.index[orders_prods_merge['prices'].isnull()])

In [31]:
orders_prods_merge.isnull().sum()

order_id                        0
user_id                         0
eval_set                        0
order_number                    0
order_dow                       0
order_hour_of_day               0
days_since_prior_order    2076096
product_id                      0
add_to_cart_order               0
reordered                       0
product_name                    0
aisle_id                        0
department_id                   0
prices                          0
_merge                          0
price_range_loc                 0
dtype: int64

In [32]:
orders_prods_merge['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

# Finally I have it lining up without any "bad" data. My data is the same as the exercise. I just have a high product range but this is within the norm. 

In [33]:
# For-loop for busiest days 

orders_prods_merge['order_dow'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_dow, dtype: int64

In [34]:
result = []

for value in orders_prods_merge["order_dow"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [35]:
print(result)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



# My notebook does not have the processing power to do such a query or the settings block it.

In [36]:
orders_prods_merge['Busiest day'] = result

In [37]:
orders_prods_merge['Busiest day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: Busiest day, dtype: int64

# TASK 4.7

# Question 2: Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [38]:
result_1 = []

for value in orders_prods_merge["order_dow"]:
  if value == 0:
    result_1.append("Busiest_days")
  elif value == 1:
    result_1.append("Busiest_days")
  elif value == 3:
    result_1.append("Slowest_days")
  elif value ==4:
    result_1.append("Slowest_days")
  else:
    result_1.append("Regularly busy")

In [39]:
result_1

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Busiest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_days',
 'Slowest_day

In [40]:
orders_prods_merge['Busiest_day'] = result_1

In [41]:
orders_prods_merge['Busiest_day'].value_counts(dropna = False)

Regularly busy    12916111
Busiest_days      11864412
Slowest_days       7624336
Name: Busiest_day, dtype: int64

## Answer: Naturally when we increase the number of days for both the busiest and slowest the numbers increase and the regular busy days decrease compared to that what we had in our exercise. 

# Question 4: When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [42]:
orders_prods_merge['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [44]:
result_2 = []


for value in orders_prods_merge["order_hour_of_day"]:
  if value > 9 & value <=17:
    result_2.append("Most_orders")
  elif value <=5:
    result_2.append("Fewest_orders")
  else:
    result_2.append("Average_orders")
  

In [45]:
result_2

['Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',

In [46]:
orders_prods_merge['busiest_period_of_day'] = result_2

# Question 5: Print the frequency for this new column.

In [47]:
orders_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

Most_orders       27898069
Average_orders     4172321
Fewest_orders       334469
Name: busiest_period_of_day, dtype: int64

In [48]:
orders_prods_merge.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                     NaN     14084.0                2.0        0.0   
2                     NaN     12427.0                3.0        0.0   
3                     NaN     26088.0                4.0        0.0   
4                     NaN     26405.0                5.0        0.0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda      77.0            7.0     9.0   
1  Organic Unsweetened Vanilla Almond Milk      91.0           16.0    12.5   
2                      Original Beef Jerky      23.0           19.0     4.4   
3               Aged White Cheddar Popcorn      23.0           19.0     4.7   
4         XL Pick-A-Size Paper Towel Rolls      54.0           17.0     1.0   

  _merge    price_range_loc     Busiest day     Busiest_day  \
0   both  Mid-range product  Regularly busy  Regularly busy   
1   both  Mid-range product  Regularly busy  Regularly busy   
2   both  Low-range product  Regularly busy  Regularly busy   
3   both  Low-range product  Regularly busy  Regularly busy   
4   both  Low-range product  Regularly busy  Regularly busy   

  busiest_period_of_day  
0        Average_orders  
1        Average_orders  
2        Average_orders  
3        Average_orders  
4        Average_orders

In [49]:
orders_prods_merge.describe()

order_id       user_id  order_number     order_dow  \
count  3.240486e+07  3.240486e+07  3.240486e+07  3.240486e+07   
mean   1.710745e+06  1.029372e+05  1.714230e+01  2.738867e+00   
std    9.872988e+05  5.946610e+04  1.753532e+01  2.090077e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.559470e+05  5.142200e+04  5.000000e+00  1.000000e+00   
50%    1.711049e+06  1.026160e+05  1.100000e+01  3.000000e+00   
75%    2.565499e+06  1.543890e+05  2.400000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01  6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.240486e+07            3.032876e+07  3.240486e+07   
mean        1.342515e+01            1.110408e+01  2.559866e+04   
std         4.246380e+00            8.779064e+00  1.408400e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            5.000000e+00  1.354400e+04   
50%         1.300000e+01            8.000000e+00  2.530200e+04   
75%         1.600000e+01            1.500000e+01  3.794700e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   

       add_to_cart_order     reordered      aisle_id  department_id  \
count       3.240486e+07  3.240486e+07  3.240486e+07   3.240486e+07   
mean        8.352547e+00  5.895873e-01  7.119612e+01   9.919792e+00   
std         7.127071e+00  4.919087e-01  3.821139e+01   6.281485e+00   
min         1.000000e+00  0.000000e+00  1.000000e+00   1.000000e+00   
25%         3.000000e+00  0.000000e+00  3.100000e+01   4.000000e+00   
50%         6.000000e+00  1.000000e+00  8.300000e+01   9.000000e+00   
75%         1.100000e+01  1.000000e+00  1.070000e+02   1.600000e+01   
max         1.450000e+02  1.000000e+00  1.340000e+02   2.100000e+01   

             prices  
count  3.240486e+07  
mean   1.198023e+01  
std    4.956554e+02  
min    1.000000e+00  
25%    4.200000e+00  
50%    7.400000e+00  
75%    1.130000e+01  
max    9.999900e+04

In [50]:
# Export file as pickle file

orders_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))